src/rust/engine/src/watch_tests.rs

```diff
@@ -0,0 +1,53 @@
+use std::sync::Arc;
+use std::path::{PathBuf};
+use std::thread::sleep;
+use std::time::Duration;
+use crate::watch::InvalidationWatcher;
+use graph::{Graph, test_support::TestGraph};
+use hashing::EMPTY_DIGEST;
+use graph::entry::{RunToken, Generation, EntryState, EntryResult};
+use crate::nodes::{DigestFile, NodeKey, NodeResult};
+use fs::File;
+use testutil::make_file;
+
+#[test]
```

Comment:
It would be good to see unit tests here also for:
* `build_root/dir/subdir/file` changing only invalidating `build_root/dir/subdir/file`
* `build_root/dir/subdir/file` being created or deleted invalidating `build_root/dir/subdir/file` and `build_root/dir/subdir`
* A symlink target being changed invalidating exactly the symlink
* For `build_root/dir/subdir/file`, `subdir` being deleted invalidating `build_root/dir`, `build_root/dir/subdir` and `build_root/dir/subdir/file`
* build_root being a symlink
* Changing the executable bit on a file invalidating that file

Hopefully these are all nicely extractable into common code so that each test is just:
* Set up a filesystem
* Call a function
* Modify the filesystem
* Make an assertion

Reply:
There will be at least one followup patch in this series: I think investing more time in testing there would be reasonable. But I also think that some of those tests should likely be in other modules: I could imagine that this module should only be testing "point" invalidation (since that is all that it actually implements), and then having a separate module that tests `impl VFS for Context`.